src/dstack/_internal/server/services/logs/gcp.py

```diff
@@ -78,35 +79,35 @@ def poll_logs(self, project: ProjectModel, request: PollLogsRequest) -> JobSubmi
             log_filters.append(f'timestamp < "{request.end_time.isoformat()}"')
         log_filter = " AND ".join(log_filters)
 
-        order_by = logging.DESCENDING if request.descending else logging.ASCENDING
+        order_by = logging_v2.DESCENDING if request.descending else logging_v2.ASCENDING
         try:
-            entries: Iterable[logging.LogEntry] = self.logger.list_entries(
-                filter_=log_filter,
+            # Use low-level API to get access to next_page_token
+            request_obj = ListLogEntriesRequest(
+                resource_names=[f"projects/{self.client.project}"],
+                filter=log_filter,
                 order_by=order_by,
-                max_results=request.limit,
-                # Specify max possible page_size (<=1000) to reduce number of API calls.
                 page_size=request.limit,
+                page_token=request.next_token,
             )
+            response = self.client._logging_api._gapic_api.list_log_entries(request=request_obj)
+
             logs = [
                 LogEvent(
                     timestamp=entry.timestamp,
-                    message=entry.payload["message"],
+                    message=entry.json_payload.get("message"),
                     log_source=LogEventSource.STDOUT,
                 )
-                for entry in entries
+                for entry in response.entries
             ]
+            next_token = response.next_page_token or None
         except google.api_core.exceptions.ResourceExhausted as e:
             logger.warning("GCP Logging exception: %s", repr(e))
             # GCP Logging has severely low quota of 60 reads/min for entries.list
             raise ServerClientError(
                 "GCP Logging read request limit exceeded."
                 " It's recommended to increase default entries.list request quota from 60 per minute."
             )
-        # We intentionally make reading logs slow to prevent hitting GCP quota.
-        # This doesn't help with many concurrent clients but
-        # should help with one client reading all logs sequentially.
-        time.sleep(1)
```

Comment:
Without it, you'd hit quota and wouldn't be able to read the logs at all.

Reply:
Yeah. Do you think we can raise a specific exception and handle it on the client?

Reply:
We already return an error when the quota is hit, see above. At that point, the client can do very little – you may have to wait for a minute or so. So we prevent quota hit in the first place.